File: IrlCPC_2018/p5.cpp
#include<bits/stdc++.h>
using namespace std;
typedef long long ll;

ll N;
ll a[8];
set<ll> sets[100];

int main(){
	scanf("%lld",&N);
	for(ll i = 0; i < N; ++i)
		scanf("%lld",&a[i]);
	sets[0].insert(0LL);
	for(ll i = 0; i < N; ++i){
		sets[1<<i].insert(a[i]);
	}
	bool survive = true;
	while(survive){
		survive = false;
		for(ll i = 0; i < (1<<N); ++i){
			for(ll j = 0; j < (1<<N); ++j){
				if((i&j) > 0)//non-disjoint
					continue;
				ll ns = i | j;
				for(ll a : sets[i]){
					for(ll b : sets[j]){
						if(sets[ns].count(a+b) <= 0){
							sets[ns].insert(a+b);
							survive = true;
						}
						if(sets[ns].count(a-b) <= 0){
							sets[ns].insert(a-b);
							survive = true;
						}
						if(sets[ns].count(a*b) <= 0){
							sets[ns].insert(a*b);
							survive = true;
						}
						if(b!=0 && a%b==0 && sets[ns].count(a/b) <= 0){
							sets[ns].insert(a/b);
							survive = true;
						}
					}
				}
			}
		}
	}
	set<ll> all;
	for(ll i = 0; i < (1<<N); ++i)
		for(ll a : sets[i])
			all.insert(a);
	ll ans = 0;
	for(ll i = 1;;++i){
		if(all.count((ll)i) <= 0){
			ans = i;
			break;
		}
	}
	printf("%lld\n",ans);
	return 0;
}

File: IrlCPC_2018/p5.py
a = [0 for _ in range(0)]
sets = [set() for _ in range(100)]

N = int(input())
a = list(map(int,input().split()))
sets[0].add(0);
for i in range(N):
    sets[1<<i].add(a[i]);
survive = True
while(survive):
        survive = False;
        for i in range(1<<N):
            for j in range(1<<N):
                    if((i&j) > 0):#non-disjoint
                            continue;
                    ns = i | j;
                    nset = set()
                    for a in sets[i]:
                        for b in sets[j]:
                                if(a+b not in sets[ns]):
                                        nset.add(a+b);
                                        survive = True;
                                if(a >= b and a-b not in sets[ns]):
                                        nset.add(a-b);
                                        survive = True;
                                if(a*b not in sets[ns]):
                                        nset.add(a*b);
                                        survive = True;
                                if(b!= 0 and a%b==0 and a//b not in sets[ns]):
                                        nset.add(a//b);
                                        survive = True;
                    sets[ns] = sets[ns].union(nset)

a = set()
for i in range(1<<N):
    for b in sets[i]:
        a.add(b)
ans = 1;
while True:
    if ans not in a:
        break
    ans +=1
print(ans)
